File: src/main/java/dev/kensa/java/WithAssertJ.java
package dev.kensa.java;

import dev.kensa.StateExtractor;
import org.assertj.core.api.*;

import static dev.kensa.context.TestContextHolder.testContext;

public interface WithAssertJ {

    default <T> OptionalAssert<T> then(OptionalStateExtractor<T> extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default OptionalDoubleAssert then(OptionalDoubleStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default OptionalIntAssert then(OptionalIntStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default OptionalLongAssert then(OptionalLongStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default BigDecimalAssert then(BigDecimalStateExtractor extractor) {
        return testContext().then(extractor, BigDecimalAssert::new);
    }

    default BigIntegerAssert then(BigIntegerStateExtractor extractor) {
        return testContext().then(extractor, BigIntegerAssert::new);
    }

    default UriAssert then(UriStateExtractor extractor) {
        return testContext().then(extractor, UriAssert::new);
    }

    default UrlAssert then(UrlStateExtractor extractor) {
        return testContext().then(extractor, UrlAssert::new);
    }

    default BooleanAssert then(BooleanStateExtractor extractor) {
        return testContext().then(extractor, BooleanAssert::new);
    }

    default StringAssert then(StringStateExtractor extractor) {
        return testContext().then(extractor, StringAssert::new);
    }

    default ByteAssert then(ByteStateExtractor extractor) {
        return testContext().then(extractor, ByteAssert::new);
    }

    default ByteArrayAssert then(ByteArrayStateExtractor extractor) {
        return testContext().then(extractor, ByteArrayAssert::new);
    }

    default CharacterAssert then(CharacterStateExtractor extractor) {
        return testContext().then(extractor, CharacterAssert::new);
    }

    default CharArrayAssert then(CharArrayStateExtractor extractor) {
        return testContext().then(extractor, CharArrayAssert::new);
    }

    default ClassAssert then(ClassStateExtractor extractor) {
        return testContext().then(extractor, ClassAssert::new);
    }

    default DoubleAssert then(DoubleStateExtractor extractor) {
        return testContext().then(extractor, DoubleAssert::new);
    }

    default DoubleArrayAssert then(DoubleArrayStateExtractor extractor) {
        return testContext().then(extractor, DoubleArrayAssert::new);
    }

    default FileAssert then(FileStateExtractor extractor) {
        return testContext().then(extractor, FileAssert::new);
    }

    default FloatAssert then(FloatStateExtractor extractor) {
        return testContext().then(extractor, FloatAssert::new);
    }

    default FloatArrayAssert then(FloatArrayStateExtractor extractor) {
        return testContext().then(extractor, FloatArrayAssert::new);
    }

    default IntegerAssert then(IntegerStateExtractor extractor) {
        return testContext().then(extractor, IntegerAssert::new);
    }

    default IntArrayAssert then(IntArrayStateExtractor extractor) {
        return testContext().then(extractor, IntArrayAssert::new);
    }

    default LongAssert then(LongStateExtractor extractor) {
        return testContext().then(extractor, LongAssert::new);
    }

    default LongArrayAssert then(LongArrayStateExtractor extractor) {
        return testContext().then(extractor, LongArrayAssert::new);
    }

    default <T> ObjectAssert<T> then(StateExtractor<T> extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default <T> ObjectArrayAssert<T> then(ObjectArrayStateExtractor<T> extractor) {
        return testContext().then(extractor, ObjectArrayAssert::new);
    }

    default ShortAssert then(ShortStateExtractor extractor) {
        return testContext().then(extractor, ShortAssert::new);
    }

    default ShortArrayAssert then(ShortArrayStateExtractor extractor) {
        return testContext().then(extractor, ShortArrayAssert::new);
    }

    default AbstractZonedDateTimeAssert<?> then(ZonedDateTimeStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default AbstractLocalDateTimeAssert<?> then(LocalDateTimeStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default AbstractOffsetDateTimeAssert<?> then(OffsetDateTimeStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default AbstractLocalTimeAssert<?> then(LocalTimeStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default AbstractLocalDateAssert<?> then(LocalDateStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default AbstractInstantAssert<?> then(InstantStateExtractor extractor) {
        return testContext().then(extractor, AssertionsForClassTypes::assertThat);
    }

    default ThrowableAssert then(ThrowableStateExtractor extractor) {
        return testContext().then(extractor, ThrowableAssert::new);
    }

    default <T> OptionalAssert<T> and(OptionalStateExtractor<T> extractor) {
        return then(extractor);
    }

    default OptionalDoubleAssert and(OptionalDoubleStateExtractor extractor) {
        return then(extractor);
    }

    default OptionalIntAssert and(OptionalIntStateExtractor extractor) {
        return then(extractor);
    }

    default OptionalLongAssert and(OptionalLongStateExtractor extractor) {
        return then(extractor);
    }

    default BigDecimalAssert and(BigDecimalStateExtractor extractor) {
        return then(extractor);
    }

    default BigIntegerAssert and(BigIntegerStateExtractor extractor) {
        return then(extractor);
    }

    default UriAssert and(UriStateExtractor extractor) {
        return then(extractor);
    }

    default UrlAssert and(UrlStateExtractor extractor) {
        return then(extractor);
    }

    default BooleanAssert and(BooleanStateExtractor extractor) {
        return then(extractor);
    }

    default StringAssert and(StringStateExtractor extractor) {
        return then(extractor);
    }

    default ByteAssert and(ByteStateExtractor extractor) {
        return then(extractor);
    }

    default ByteArrayAssert and(ByteArrayStateExtractor extractor) {
        return then(extractor);
    }

    default CharacterAssert and(CharacterStateExtractor extractor) {
        return then(extractor);
    }

    default CharArrayAssert and(CharArrayStateExtractor extractor) {
        return then(extractor);
    }

    default ClassAssert and(ClassStateExtractor extractor) {
        return then(extractor);
    }

    default DoubleAssert and(DoubleStateExtractor extractor) {
        return then(extractor);
    }

    default DoubleArrayAssert and(DoubleArrayStateExtractor extractor) {
        return then(extractor);
    }

    default FileAssert and(FileStateExtractor extractor) {
        return then(extractor);
    }

    default FloatAssert and(FloatStateExtractor extractor) {
        return then(extractor);
    }

    default FloatArrayAssert and(FloatArrayStateExtractor extractor) {
        return then(extractor);
    }

    default IntegerAssert and(IntegerStateExtractor extractor) {
        return then(extractor);
    }

    default IntArrayAssert and(IntArrayStateExtractor extractor) {
        return then(extractor);
    }

    default LongAssert and(LongStateExtractor extractor) {
        return then(extractor);
    }

    default LongArrayAssert and(LongArrayStateExtractor extractor) {
        return then(extractor);
    }

    default <T> ObjectAssert<T> and(StateExtractor<T> extractor) {
        return then(extractor);
    }

    default <T> ObjectArrayAssert<T> and(ObjectArrayStateExtractor<T> extractor) {
        return then(extractor);
    }

    default ShortAssert and(ShortStateExtractor extractor) {
        return then(extractor);
    }

    default ShortArrayAssert and(ShortArrayStateExtractor extractor) {
        return then(extractor);
    }

    default AbstractZonedDateTimeAssert<?> and(ZonedDateTimeStateExtractor extractor) {
        return then(extractor);
    }

    default AbstractLocalDateTimeAssert<?> and(LocalDateTimeStateExtractor extractor) {
        return then(extractor);
    }

    default AbstractOffsetDateTimeAssert<?> and(OffsetDateTimeStateExtractor extractor) {
        return then(extractor);
    }

    default AbstractLocalTimeAssert<?> and(LocalTimeStateExtractor extractor) {
        return then(extractor);
    }

    default AbstractLocalDateAssert<?> and(LocalDateStateExtractor extractor) {
        return then(extractor);
    }

    default AbstractInstantAssert<?> and(InstantStateExtractor extractor) {
        return then(extractor);
    }

    default ThrowableAssert and(ThrowableStateExtractor extractor) {
        return then(extractor);
    }
}

File: src/main/java/dev/kensa/kotlin/WithAssertJ.kt
package dev.kensa.kotlin

import dev.kensa.StateExtractor
import dev.kensa.context.AssertJThen
import dev.kensa.context.TestContextHolder.testContext
import org.assertj.core.api.*
import java.io.File
import java.math.BigDecimal
import java.math.BigInteger
import java.net.URI
import java.net.URL
import java.time.*
import java.util.*
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

interface WithAssertJ {

    private fun <T, A> then(extractor: StateExtractor<T>, assertProvider: (T?) -> A): A =
        AssertJThen.then(testContext(), extractor, assertProvider)

    fun <T, A> thenEventually(duration: Duration = 10.seconds, extractor: StateExtractor<T>, assertProvider: (T?) -> A, block: A.() -> Unit) =
        AssertJThen.thenEventually(duration, testContext(), extractor) { actual ->
            assertProvider(actual).apply(block)
        }

    fun <T, A> then(extractor: StateExtractor<T>, assertProvider: (T?) -> A, block: A.() -> Unit) =
        AssertJThen.then(testContext(), extractor) { actual ->
            assertProvider(actual).apply(block)
        }

    fun <T> then(extractor: StateExtractor<Optional<T>>): OptionalAssert<T> = then(extractor) { actual: Optional<T>? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<OptionalDouble>): OptionalDoubleAssert = then(extractor) { actual: OptionalDouble? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<OptionalInt>): OptionalIntAssert = then(extractor) { actual: OptionalInt? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<OptionalLong>): OptionalLongAssert = then(extractor) { actual: OptionalLong? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<BigDecimal?>): BigDecimalAssert = then(extractor, ::BigDecimalAssert)

    fun then(extractor: StateExtractor<BigInteger?>): BigIntegerAssert = then(extractor, ::BigIntegerAssert)

    fun then(extractor: StateExtractor<URI?>): UriAssert = then(extractor, ::UriAssert)

    fun then(extractor: StateExtractor<URL?>): UrlAssert = then(extractor, ::UrlAssert)

    fun then(extractor: StateExtractor<Boolean?>): BooleanAssert = then(extractor, ::BooleanAssert)

    fun then(extractor: StateExtractor<String?>): StringAssert = then(extractor, ::StringAssert)

    fun then(extractor: StateExtractor<Byte?>): ByteAssert = then(extractor, ::ByteAssert)

    fun then(extractor: StateExtractor<ByteArray?>): ByteArrayAssert = then(extractor, ::ByteArrayAssert)

    fun then(extractor: StateExtractor<Char?>): CharacterAssert = then(extractor, ::CharacterAssert)

    fun then(extractor: StateExtractor<CharArray?>): CharArrayAssert = then(extractor, ::CharArrayAssert)

    fun then(extractor: StateExtractor<Class<*>?>): ClassAssert = then(extractor, ::ClassAssert)

    fun then(extractor: StateExtractor<Double?>): DoubleAssert = then(extractor) { actual: Double? -> DoubleAssert(actual) }

    fun then(extractor: StateExtractor<DoubleArray?>): DoubleArrayAssert = then(extractor, ::DoubleArrayAssert)

    fun then(extractor: StateExtractor<File?>): FileAssert = then(extractor, ::FileAssert)

    fun then(extractor: StateExtractor<Float?>): FloatAssert = then(extractor) { actual: Float? -> FloatAssert(actual) }

    fun then(extractor: StateExtractor<FloatArray?>): FloatArrayAssert = then(extractor, ::FloatArrayAssert)

    fun then(extractor: StateExtractor<Int?>): IntegerAssert = then(extractor, ::IntegerAssert)

    fun then(extractor: StateExtractor<IntArray?>): IntArrayAssert = then(extractor, ::IntArrayAssert)

    fun then(extractor: StateExtractor<Long?>): LongAssert = then(extractor, ::LongAssert)

    fun then(extractor: StateExtractor<LongArray?>): LongArrayAssert = then(extractor, ::LongArrayAssert)

    fun then(extractor: StateExtractor<Any?>): ObjectAssert<*> = then(extractor) { actual: Any? -> ObjectAssert(actual) }

    fun then(extractor: StateExtractor<Array<Any>?>): ObjectArrayAssert<*> = then(extractor) { actual: Array<Any>? -> ObjectArrayAssert(actual) }

    fun then(extractor: StateExtractor<Short?>): ShortAssert = then(extractor, ::ShortAssert)

    fun then(extractor: StateExtractor<ShortArray?>): ShortArrayAssert = then(extractor, ::ShortArrayAssert)

    fun then(extractor: StateExtractor<ZonedDateTime?>): AbstractZonedDateTimeAssert<*> = then(extractor) { actual: ZonedDateTime? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<LocalDateTime?>): AbstractLocalDateTimeAssert<*> = then(extractor) { actual: LocalDateTime? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<OffsetDateTime?>): AbstractOffsetDateTimeAssert<*> = then(extractor) { actual: OffsetDateTime? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<LocalTime?>): AbstractLocalTimeAssert<*> = then(extractor) { actual: LocalTime? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<LocalDate?>): AbstractLocalDateAssert<*> = then(extractor) { actual: LocalDate? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<Instant?>): AbstractInstantAssert<*> = then(extractor) { actual: Instant? -> AssertionsForClassTypes.assertThat(actual) }

    fun then(extractor: StateExtractor<Throwable?>): ThrowableAssert<*> = then(extractor, ::ThrowableAssert)

    fun <T> and(extractor: StateExtractor<Optional<T>>): OptionalAssert<T> = then(extractor)

    fun and(extractor: StateExtractor<OptionalDouble>): OptionalDoubleAssert = then(extractor)

    fun and(extractor: StateExtractor<OptionalInt>): OptionalIntAssert = then(extractor)

    fun and(extractor: StateExtractor<OptionalLong>): OptionalLongAssert = then(extractor)

    fun and(extractor: StateExtractor<BigDecimal?>): BigDecimalAssert = then(extractor)

    fun and(extractor: StateExtractor<BigInteger?>): BigIntegerAssert = then(extractor)

    fun and(extractor: StateExtractor<URI?>): UriAssert = then(extractor)

    fun and(extractor: StateExtractor<URL?>): UrlAssert = then(extractor)

    fun and(extractor: StateExtractor<Boolean?>): BooleanAssert = then(extractor)

    fun and(extractor: StateExtractor<String?>): StringAssert = then(extractor)

    fun and(extractor: StateExtractor<Byte?>): ByteAssert = then(extractor)

    fun and(extractor: StateExtractor<ByteArray?>): ByteArrayAssert = then(extractor)

    fun and(extractor: StateExtractor<Char?>): CharacterAssert = then(extractor)

    fun and(extractor: StateExtractor<CharArray?>): CharArrayAssert = then(extractor)

    fun and(extractor: StateExtractor<Class<*>?>): ClassAssert = then(extractor)

    fun and(extractor: StateExtractor<Double?>): DoubleAssert = then(extractor)

    fun and(extractor: StateExtractor<DoubleArray?>): DoubleArrayAssert = then(extractor)

    fun and(extractor: StateExtractor<File?>): FileAssert = then(extractor)

    fun and(extractor: StateExtractor<Float?>): FloatAssert = then(extractor)

    fun and(extractor: StateExtractor<FloatArray?>): FloatArrayAssert = then(extractor)

    fun and(extractor: StateExtractor<Int?>): IntegerAssert = then(extractor)

    fun and(extractor: StateExtractor<IntArray?>): IntArrayAssert = then(extractor)

    fun and(extractor: StateExtractor<Long?>): LongAssert = then(extractor)

    fun and(extractor: StateExtractor<LongArray?>): LongArrayAssert = then(extractor)

    fun and(extractor: StateExtractor<Any?>): ObjectAssert<*> = then(extractor)

    fun and(extractor: StateExtractor<Array<Any>?>): ObjectArrayAssert<*> = then(extractor)

    fun and(extractor: StateExtractor<Short?>): ShortAssert = then(extractor)

    fun and(extractor: StateExtractor<ShortArray?>): ShortArrayAssert = then(extractor)

    fun and(extractor: StateExtractor<ZonedDateTime?>): AbstractZonedDateTimeAssert<*> = then(extractor)

    fun and(extractor: StateExtractor<LocalDateTime?>): AbstractLocalDateTimeAssert<*> = then(extractor)

    fun and(extractor: StateExtractor<OffsetDateTime?>): AbstractOffsetDateTimeAssert<*> = then(extractor)

    fun and(extractor: StateExtractor<LocalTime?>): AbstractLocalTimeAssert<*> = then(extractor)

    fun and(extractor: StateExtractor<LocalDate?>): AbstractLocalDateAssert<*> = then(extractor)

    fun and(extractor: StateExtractor<Instant?>): AbstractInstantAssert<*> = then(extractor)

    fun and(extractor: StateExtractor<Throwable?>): ThrowableAssert<*> = then(extractor)

}